cylc/flow/rundb.py

```diff
@@ -663,19 +663,19 @@ def select_submit_nums_for_insert(self, task_ids):
         task_ids should be specified as [(name-glob, cycle), ...]
 
         """
-        stmt = r"SELECT name,cycle,submit_num FROM %(name)s" % {
-            "name": self.TABLE_TASK_STATES}
-        stmt_args = []
-        if task_ids:
-            stmt += (
-                " WHERE (" +
-                ") OR (".join(["name GLOB ? AND cycle==?"] * len(task_ids)) +
-                ")")
-            for name, cycle in task_ids:
-                stmt_args += [name, cycle]
+        # Ignore bandit false positive: B608: hardcoded_sql_expressions
+        # Not an injection, simply putting the table name in the SQL query
+        # expression as a string constant local to this module.
+        stmt = (  # nosec
+            r"SELECT name,cycle,submit_num FROM %(name)s"
+            r" WHERE name==? AND CYCLE==?"
+        ) % {"name": self.TABLE_TASK_STATES}
         ret = {}
-        for name, cycle, submit_num in self.connect().execute(stmt, stmt_args):
-            ret[(name, cycle)] = submit_num
+        for name, cycle in task_ids:
+            for name, cycle, submit_num in self.connect().execute(
```

Comment:
Out of interest, why move this loop out of SQL?

Reply:
This was prompted by a test failure because it has too many WHERE clauses. So...

Instead of building a single statement with multiple WHERE clauses, execute statement multiple times to avoid issues when there are too many WHERE clauses. This may be less efficient, but is much safer.